File: bindings/nodejs/lib/session.js
const native = require("./native");
const { Result } = require("./result");

class Session {
    constructor(handle) {
        this.handle = handle;
    }

    reset() {
        native.fn.vietasr_session_reset(this.handle);
    }

    accept(pcm) {
        if (!this.handle) {
            throw new Error("Session is closed");
        }
        if (pcm instanceof Int16Array) {
            return native.fn.vietasr_accept_waveform_s16(
                this.handle, pcm, pcm.length) === 1;
        }
        if (pcm instanceof Float32Array) {
            return native.fn.vietasr_accept_waveform_f32(
                this.handle, pcm, pcm.length) === 1;
        }
        if (Buffer.isBuffer(pcm)) {
            const view = new Int16Array(pcm.buffer, pcm.byteOffset,
                pcm.byteLength / 2);
            return native.fn.vietasr_accept_waveform_s16(
                this.handle, view, view.length) === 1;
        }
        throw new TypeError(
            "accept() expects Int16Array, Float32Array, or Buffer (16-bit PCM)");
    }

    partial() {
        return Result.fromJson(native.fn.vietasr_partial_result(this.handle));
    }

    result() {
        return Result.fromJson(native.fn.vietasr_result(this.handle));
    }

    final() {
        return Result.fromJson(native.fn.vietasr_final_result(this.handle));
    }

    close() {
        if (this.handle) {
            native.fn.vietasr_session_free(this.handle);
            this.handle = null;
        }
    }
}

module.exports = { Session };

File: bindings/java/src/main/java/io/vietasr/Session.java
package io.vietasr;

import com.sun.jna.Pointer;

public final class Session implements AutoCloseable {
    private Pointer handle;

    Session(Pointer handle) {
        this.handle = handle;
    }

    public void reset() {
        NativeLibrary.INSTANCE.vietasr_session_reset(handle);
    }

    public boolean accept(short[] pcm) {
        if (handle == null) {
            throw new VietasrException("Session is closed");
        }
        return NativeLibrary.INSTANCE.vietasr_accept_waveform_s16(handle, pcm, pcm.length) == 1;
    }

    public boolean accept(float[] pcm) {
        if (handle == null) {
            throw new VietasrException("Session is closed");
        }
        return NativeLibrary.INSTANCE.vietasr_accept_waveform_f32(handle, pcm, pcm.length) == 1;
    }

    public Result partial() {
        return Result.fromJson(NativeLibrary.INSTANCE.vietasr_partial_result(handle));
    }

    public Result result() {
        return Result.fromJson(NativeLibrary.INSTANCE.vietasr_result(handle));
    }

    public Result finalResult() {
        return Result.fromJson(NativeLibrary.INSTANCE.vietasr_final_result(handle));
    }

    @Override
    public void close() {
        if (handle != null) {
            NativeLibrary.INSTANCE.vietasr_session_free(handle);
            handle = null;
        }
    }
}
